File: 234-palindrome-linked-list/234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head==null||head.next==null) return true;
        
        Stack<Integer> st=new Stack();
        ListNode start=new ListNode(0);
        ListNode sp=start, fp=start;
        start.next=head;
        
        while(fp!=null && fp.next!=null){
            sp=sp.next;
            fp=fp.next.next;
            st.push(sp.val);
        }
        if(fp!=null) sp=sp.next;
    
        while(!st.isEmpty() && sp.val==st.peek()){
            sp=sp.next;
            st.pop();
        }      
        if(st.isEmpty()) return true;
        return false;
    }
}

File: 234-palindrome-linked-list/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        #find mid
        sp=fp=head
        while fp and fp.next:
            sp=sp.next
            fp=fp.next.next
        #reverse next half
        prv=None
        while sp:
            nxt=sp.next
            sp.next=prv
            prv=sp
            sp=nxt
        #compare 1st &  2nd half
        sp=head
        while prv:
            if prv.val!=head.val:
                return False
            prv=prv.next
            head=head.next
        return True
        